xds/internal/balancer/clusterresolver/configbuilder.go

```diff
@@ -198,7 +198,7 @@ func buildClusterImplConfigForDNS(g *nameGenerator, addrStrs []string, mechanism
 	retAddrs := make([]resolver.Address, 0, len(addrStrs))
 	pName := fmt.Sprintf("priority-%v", g.prefix)
 	for _, addrStr := range addrStrs {
-		retAddrs = append(retAddrs, hierarchy.Set(resolver.Address{Addr: addrStr}, []string{pName}))
+		retAddrs = append(retAddrs, hierarchy.Set(resolver.Address{Addr: addrStr, ServerName: mechanism.DNSHostname}, []string{pName}))
```

Comment:
This PR just adds this change.
Other changes are inherited from https://github.com/grpc/grpc-go/pull/5748 as I described in the description of this PR.

Reply:
Our current behavior (which is to set the `:authority` header to the authority on the channel and not the DNS name received as part of the Cluster resource) was an explicit choice. It is a fairly serious security risk to set it to the DNS name received from the Cluster resource.

We are going to be investigating this further at some point to support certain serverless use cases. But it's not yet clear to us, how we will solve the security problem here. We'll definitely need more design work before we can safely do something like this. And when we have a design for this, we will publish a gRFC for the same, and the change will happen across all supported gRPC languages.

Unfortunately, we cannot accept this PR at this point in time.

Reply:
Please feel free to engage with the gRPC team on our [mailing list](https://groups.google.com/g/grpc-io) if you want to continue this discussion and work towards a secure solution for this problem. Thank you.

Reply:
@easwars 

Thank you for your review!

> It is a fairly serious security risk to set it to the DNS name received from the Cluster resource.

Could you please let me know the risk regarding this change briefly...?

Reply:
In general, the authority for the connection is used to verify the peer -- e.g., for TLS, the client checks that the server's certificate matches that authority.  Because of that, we essentially view it as a requirement that the authority be explicitly specified by the client application (either via the target URI used to create the channel or via an explicit channel option indicating the authority to use) rather than coming from some external control plane (e.g., being specified by the resolver or by an LB policy), because otherwise we would basically give a compromised control plane the ability to tell us to talk to some arbitrary endpoint without us actually verifing that the endpoint is authenticated as the intended peer.